### PR TITLE
Add CatmullRomCurve3 which supports catmullrom, centripetal & chordal

### PR DESCRIPTION
**Pull request by zz85**
parameterizations for avoiding cusps and self-intersections in non-uniform catmull rom curves. 

Continuing from #5503.

Additional questions:
1. do we want to remove `SplineCurve3` (and replace with CatmullRomCurve3 in examples)?
2. name the uniform `CatmullRomCurve3.type` as `uniform` or `catmullrom`?
3. default CatmullRomCurve3 to uniform or centripetal type?

### CODE REVIEW

**Comment by WestLangley** [comment]
It seems `SplineCurve3` and  `uniform CatmullRomCurve3`can cause unwanted cusps (unless tension is zero, in which case the latter appears to reduce to just a piecewise-linear line). I am thinking they should be removed...

That leaves `centripetal` and `chordal`. Keep both?

**Comment by zz85** [comment]
my take is that we should use `centripetal` as default going forward. however i think the uniform catmull-rom (and the tension parameter) is quite common (used in other libraries) and i'm not sure the impact of removing it completely. any thoughts @mrdoob?

**Comment by mrdoob** [comment]
This goes a bit over my head at the moment. I let you guys decide... :)

**Comment by WestLangley** [comment]
> my take is that we should use centripetal as default going forward.

That sounds fine to me. I will go along with whatever you think is best, @zz85.

Adding your spline editor as an example would be great, if that is OK with you.

**Comment by zz85** [comment]
@WestLangley - yeah i could add the example.

So i guess I could go with one of these 2 approaches
1. Support catmullrom, centripetal & chordal in `CatmullRomCurve3` and place a depreciation message in `SplineCurve3` on the migration.
2. Leave `SplineCurve3` as it is, support only centripetal & chordal in `CatmullRomCurve3`.

**Comment by WestLangley** [comment]
I believe 1 is best. It would be helpful if the `type` descriptions were a bit more detailed, so users have a clearer idea of what type to specify... Maybe that is what the docs are for. : - )

I do find your spline editor very helpful. Thank you!

**Comment by zz85** [comment]
Pull Request updated.

File: examples/js/controls/DragControls.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+		
+		}
+
+		return me;
+
+	};
+
+	var notify = function( event, data, member ) {
+		
+var l = _listeners[ event ];
```
Comment (mrdoob) [inline examples/js/controls/DragControls.js]: Seems like the formatting in this function is broken?

Comment (zz85) [inline examples/js/controls/DragControls.js]: oh yes... might be a bug in the formatter. fixed!

**Comment by zz85** [comment]
@mrdoob any updates on this?

File: src/extras/curves/SplineCurve3.js [hunk 2]
```diff
@@ -7,6 +7,7 @@ THREE.SplineCurve3 = THREE.Curve.create(
 
 	function ( points /* array of Vector3 */) {
 
+		THREE.warn( 'THREE.SplineCurve3 will be deprecated. Please use THREE.CatmullRomCurve3' );
```
Comment (mrdoob) [inline src/extras/curves/SplineCurve3.js]: Can we deprecate (remove the code) already?

Comment (zz85) [inline src/extras/curves/SplineCurve3.js]: If you intent for users to get an error while using SplineCurve3 after this change, then yes. Would you like to remove the code now?

Comment (mrdoob) [inline src/extras/curves/SplineCurve3.js]: Oh! I see, I see. Never mind :)

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**